CONTRIBUTING.md

```diff
@@ -1,30 +1,241 @@
+
+impl Serialize for SerdeThing { /* ... */ }
+impl DeserializeOwned for SerdeThing { /* ... */ }
+```
+
+If there _is_ such an advantage, then it's usually better to wrap the types. If there isn't an advantage, then you should re-export any types you're using (or expecting your users to use).
+
+Don't:
+
+```rust
+use other_crate::ExternalThing;
+
+pub fn my_function(thing: ExternalThing) -> bool {
+    /* ... */
+}
+```
+
+Do:
+
+```rust
+pub use other_crate::ExternalThing;
+
+pub fn my_function(thing: ExternalThing) -> bool {
+    /* ... */
+}
+```
+
+### Export types at the crate level
+
+The Rust Style Guide contains the admonition to [Reexport the most important types at the crate level](https://doc.rust-lang.org/1.0.0/style/style/organization.html#reexport-the-most-important-types-at-the-crate-level.). We should take this a step beyond, and simply re-export *all* publicly visible types at the crate level.
+
+Don't:
+
+```rust
+pub use crate::module::SomeType;
+
+mod module {
+    pub struct SomeType;
+
+    // Users will have to access this using `thecrate::module::Error`?
+    pub enum Error {
+        Error1,
+        Error2,
+    }
+
+    enum Strategy {
+        Strat1(SomeType),
+        Strat2,
+    }
+}
+```
+
+Do:
+
+```rust
+pub use crate::module::{SomeType, Error as SomeError};
+
+mod module {
+    pub struct SomeType;
+    pub enum Error {
+        Error1,
+        Error2,
+    }
+
+    enum Strategy {
+        Strat1(SomeType),
+        Strat2,
+    }
+}
+```
+
+### Use public modules to group functions
+
+This is the flip side of the Rust style guide's admonition to [prefer fully importing types/traits while module-qualifying functions](https://doc.rust-lang.org/1.0.0/style/style/imports.html#prefer-fully-importing-types/traits-while-module-qualifying-functions.) and our own admonition to [export types at the crate level](#export-types-at-the-crate-level).
+
+Types should only be exported at the top level, but if you have lots of related bare functions, consider grouping them into modules exposed via a `pub mod`.
+
+Don't:
+
+```rust
+pub struct SomeType;
+
+pub fn foonary_frobnicate() -> bool {
+    /* ... */
+}
+
+pub fn foonary_widgify() -> SomeType {
+    /* ... */
+}
+```
+
+Don't:
+
+```rust
+mod foonary {
+    pub struct SomeType;
+
+    pub fn frobnicate() -> SomeType {
+        /* ... */
+    }
+
+    pub fn widgify() -> SomeType {
+        /* ... */
+    }
+}
+
+pub use crate::foonary::{
+    SomeType,
+    frobnicate as foonary_frobnicate,
+    widgify as foonary_widgify,
+};
+```
+
+Do:
+
+```rust
+pub use crate::foonary::SomeType;
+
+pub mod foonary {
+    pub(crate) struct SomeType;
+
+    pub fn frobnicate() -> SomeType {
+        /* ... */
+    }
+
+    pub fn widgify() -> SomeType {
+        /* ... */
+    }
+}
+```
+
+### Avoid Manual Drops
+
+Don't use `core::drop()`. There are exceptions to every rule, but you should always look to use an inner scope instead of `core::drop()`.
```

Comment:
Unfortunately I don't have any suggestions for how to address, but when I came to this part I thought the document was contradicting itself.
Earlier this document calls out, https://doc.rust-lang.org/1.0.0/style/features/let.html#always-separately-bind-raii-guards.-%5Bfixme:-needs-rfc%5D, which starts out with an example utilizing `drop`.  Nested at the bottom of the paragraph is
> Consider using an explicit drop to release the resource early.

It may just be my rust newness or reading too fast, but I came away thinking it was encouraging the use of drop for RAII. (my rust newness) I didn't fully register the `let binding`, even though it's the title, I ended up interpreting it as use explicit lock and unlock based on the sample code.

Reply:
Yeah, we're recommending the same thing, ultimately, I'm just saying "yeah but use scope instead of `drop()`", e.g.:

```rust
{
    let binding = foo.lock();
    binding.do_stuff();
}

// do a bunch of non-synchronized stuff here
```

is superior to

```rust
let binding = foo.lock();
binding.do_stuff();
drop(binding);

// do a bunch of non-synchronized stuff here
```

I'll update with that example.